src/main/java/org/ambraproject/wombat/controller/ArticleController.java

```diff
@@ -806,15 +807,55 @@ private XmlContent getXmlContent(Site site, ArticlePointer articlePointer,
       byte[] xml = ByteStreams.toByteArray(stream);
       final Document document = parseXmlService.getDocument(new ByteArrayInputStream(xml));
 
-      List<Reference> references = parseXmlService.parseArticleReferences(document,
-          doi -> getLinkText(site, request, doi));
+      long time1 = System.currentTimeMillis();
+
+      // do not supply Solr related link service now
+      List<Reference> references = parseXmlService.parseArticleReferences(document, null);
+
+      long time2 = System.currentTimeMillis();
+
+      // invoke one Solr API to resolve all references dois to journal keys
+      List<String> dois = references.stream().map(ref -> {
+        return ref.getDoi();
+      }).filter(doi -> {
+        return doi != null && doi.startsWith("10.1371/");
+      }). collect(Collectors.toList());
+
+      List<String> keys = doiToJournalResolutionService.getJournalKeysFromDois(dois, site);
+
+      long time3 = System.currentTimeMillis();
+
+      // store the link text from journal key to references.
+      // since Reference is immutable, need to create a new list of new reference objects.
+      Iterator<Reference> itRef = references.iterator();
+      Iterator<String> itKey = keys.iterator();
+      List<Reference> referencesWithLinks = new ArrayList<Reference>();
+      while (itRef.hasNext()) {
+        Reference ref = itRef.next();
+        if (ref.getDoi() != null && ref.getDoi().startsWith("10.1371/")) {
+          String key = itKey.next();
+          Reference.Builder builder = new Reference.Builder(ref);
+          Reference refWithLink = builder.setFullArticleLink(getLinkText(site, request, ref.getDoi(), key)).build();
+          referencesWithLinks.add(refWithLink);
+        } else {
+          referencesWithLinks.add(ref);
+        }
+      }
+      references = referencesWithLinks;
+
+      long time4 = System.currentTimeMillis();
 
       StringWriter articleHtml = new StringWriter(XFORM_BUFFER_SIZE);
       try (OutputStream outputStream = new WriterOutputStream(articleHtml, charset)) {
         articleTransformService.transformArticle(site, articlePointer, references,
             new ByteArrayInputStream(xml), outputStream);
       }
 
+      long time5 = System.currentTimeMillis();
+      System.out.println("total time=" + (time5 - time1) + " transform=" + (time5 - time4)
+          + " references=" + (time4 - time1) + " including solr=" + (time3 - time2));
+
```

Comment:
I would like to keep these println as easier for testing. Once testing done, I can remove that. The last solr time is what is being optimized here. Previously, depending on the number and type of references it could be few seconds. Now it should be less than a second, preferably less than 200 millisecond. I tested with article journal.pone.0127757